controllers/storagecluster/onboarding_validation_keys_generator_job.go

```diff
@@ -0,0 +1,114 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/util"
+)
+
+const (
+	providerAPIServerImage              = "PROVIDER_API_SERVER_IMAGE"
+	onboardingSecretGeneratorImage      = "ONBOARDING_SECRET_GENERATOR_IMAGE"
+	onboardingJobName                   = "onboarding-secret-generator"
+	onboardingTicketPublicKeySecretName = "onboarding-ticket-key"
+)
+
+type onboardingValidationKeysGeneratorJob struct{}
+
+func (o *onboardingValidationKeysGeneratorJob) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
```

Comment:
I'm just thinking whether we are successful if we created a job or ensured public keys 🤔? I see two issues:

1. Every successful reconcile basically rotates keys and if we are failing in some other reconciler after this, then we are potentially rotating keys until storagecluster stabilizes, even with that due to some spurious reconciles we will potentially be rerunning this job every minute, is that a good thing?
2. Assuming everything is stable and we aren't getting any reconciles (most ideal case) how do we cope up from job failed to create/rotate keys?

I don't think these questions are particularly related to the refactor, they just came up since we are using this feature from some time.

Unlike the status reporter job at client side, w/ current job we are closely coupled w/ outcome of the job, isn't it?

Reply:
At this location, we are successful if we create the job, While the job responsibility is to rotate the key.

Regarding 1. We are not rotating the key every time. Instead, we are checking for the existence of the public key. Therefore, even if it reconciles, the public key will exist the next time, and a job won't be created.
Regarding 2. I don't see a reason for the job to fail. If it did fail, there must be something wrong with the environment, and we need to address it.

Reply:
@leelavg The key point is point #1 in @rchikatw's answer. We will not rotate the key if even one reconcile already executed this code

Reply:
yes, thanks for clarification, can be resolved.